src/main/kotlin/org/rust/lang/core/grammar/rust.bnf

```diff
@@ -365,18 +366,45 @@ private block_item_group ::= &(CONST? UNSAFE? extern_abi? FN) fn_item
 // Functions
 ///////////////////////////////////////////////////////////////////////////////////////////////////
 
+// Parameters
 // TODO: Maybe enforce @NotNull's in specific cases
 fake parameter  ::= pat? type?
-fake parameters ::= self_argument? parameter*
+
+fn_parameter ::= pat type_ascription
+lambda_parameter ::= pat type_ascription?
+anon_parameter ::= [ restricted_pattern ':' ] type
+path_parameter ::= type !'='
+
+fake parameters          ::= self_argument? parameter*
+fake variadic_parameters ::= self_argument? parameter* '...'? { extends = parameters }
+
+fn_parameters          ::= '(' !',' [ <<comma_separated_list fn_parameter>> ] ')' { pin = 1 }
+lambda_parameters      ::= '|' !',' [ <<comma_separated_list lambda_parameter>> ] '|'
```

Comment:
I believe there is no need for this `!','` any more. They were introduced in https://github.com/intellij-rust/intellij-rust/commit/7513af403162f656dee3fa7ea5174986a4e0733a and it was not a great idea I think :)